File: src/com/github/unchama/seichiassist/commands/mebiusCommand.java
package com.github.unchama.seichiassist.commands;

import java.util.List;

import com.github.unchama.seichiassist.data.*;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabExecutor;
import org.bukkit.entity.Player;

import com.github.unchama.seichiassist.SeichiAssist;
import com.github.unchama.seichiassist.listener.MebiusListener;
import sun.management.*;

public class mebiusCommand implements TabExecutor {

	public mebiusCommand(SeichiAssist plugin) {
	}

	@Override
	public List<String> onTabComplete(CommandSender sender, Command command, String label, String[] args) {
		return null;
	}

	@Override
	public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
		if (!(sender instanceof Player)) {
			showDescription(sender);
			sender.sendMessage(ChatColor.RED + "コマンドはプレイヤーから実行してください.");
			return true;
		}

		switch (args.length) {
			case 1: {
				switch (args[0]) {
					case "get":
						if (!sender.isOp()) {
							sendPermissionWarning(sender);
							break;
						}
						MebiusListener.debugGive((Player) sender);
						break;
					case "reload":
						if (!sender.isOp()) {
							sendPermissionWarning(sender);
							break;
						}
						MebiusListener.reload();
						break;
					case "debug":
						if (!sender.isOp()) {
							sendPermissionWarning(sender);
							break;
						}
						MebiusListener.debug((Player) sender);
						break;
					case "nickname":
						String name = MebiusListener.getNickname((Player) sender);
						if (name == null) {
							sender.sendMessage(ChatColor.RED + "呼び名の確認はMEBIUSを装着して行ってください.");
						} else {
							sender.sendMessage(ChatColor.GREEN + "現在のメビウスからの呼び名 : " + name);
						}
						break;

					default:
						showDescription(sender);
						break;
				}
				return true;
			}

			case 2: {
				switch (args[0]) {
					case "naming": {
						if (!MebiusListener.setName((Player) sender, args[1])) {
							sender.sendMessage(ChatColor.RED + "命名はMEBIUSを装着して行ってください.");
						}
						return true;
					}

					case "nickname": {
						if (args[1].equals("reset")) {
							if (!MebiusListener.setNickname((Player) sender, sender.getName())) {
								sender.sendMessage(ChatColor.RED + "呼び名のリセットはMEBIUSを装着して行ってください.");
								return true;
							} else {
								sender.sendMessage(ChatColor.GREEN + "メビウスからの呼び名を" + sender.getName() + "にリセットしました.");
								return true;
							}
						}

						if (!MebiusListener.setNickname((Player) sender, args[1])) {
							sender.sendMessage(ChatColor.RED + "呼び名の設定はMEBIUSを装着して行ってください.");
							return true;
						} else {
							sender.sendMessage(ChatColor.GREEN + "メビウスからの呼び名を" + args[1] + "にセットしました.");
							return true;
						}
					}

					default:
						showDescription(sender);
						return true;
				}
			}

			default:
				showDescription(sender);
				break;
		}
		return true;
	}

	private void showDescription(CommandSender sender) {
		sender.sendMessage(ChatColor.RED + "[Usage]");
		sender.sendMessage(ChatColor.RED + "/mebius naming <name>");
		sender.sendMessage(ChatColor.RED + "  現在頭に装着中のMEBIUSに<name>を命名します。");
		sender.sendMessage("");
		sender.sendMessage(ChatColor.RED + "/mebius nickname");
		sender.sendMessage(ChatColor.RED + "  MEBIUSから呼ばれる名前を表示します");
		sender.sendMessage("");
		sender.sendMessage(ChatColor.RED + "/mebius nickname <name>");
		sender.sendMessage(ChatColor.RED + "  MEBIUSから呼ばれる名前を<name>に変更します");
		sender.sendMessage("");
		sender.sendMessage(ChatColor.RED + "/mebius nickname reset");
		sender.sendMessage(ChatColor.RED + "  MEBIUS空の呼び名をプレイヤー名(初期設定)に戻します");
		sender.sendMessage("");
	}

	private void sendPermissionWarning(CommandSender sender) {
		sender.sendMessage(ChatColor.RED + "このコマンドは権限者のみが実行可能です.");
	}
}

File: src/main/scala/com/github/unchama/seichiassist/mebius/controller/command/MebiusCommand.scala
package com.github.unchama.seichiassist.mebius.controller.command

import cats.effect.IO
import com.github.unchama.contextualexecutor.ContextualExecutor
import com.github.unchama.contextualexecutor.builder.{ContextualExecutorBuilder, Parsers}
import com.github.unchama.contextualexecutor.executors.BranchedExecutor
import com.github.unchama.seichiassist.SeichiAssist
import com.github.unchama.seichiassist.commands.contextual.builder.BuilderTemplates.playerCommandBuilder
import com.github.unchama.seichiassist.mebius.controller.codec.ItemStackMebiusCodec
import com.github.unchama.seichiassist.mebius.controller.listeners.MebiusListener
import com.github.unchama.targetedeffect.TargetedEffect
import com.github.unchama.targetedeffect.TargetedEffect.emptyEffect
import com.github.unchama.targetedeffect.commandsender.MessageEffect
import org.bukkit.ChatColor.{GREEN, RED, RESET}
import org.bukkit.command.{CommandSender, TabExecutor}
import org.bukkit.entity.Player

object MebiusCommand {

  import ChildExecutors._

  val executor: TabExecutor = {

    BranchedExecutor(
      Map(
        "nickname" -> ChildExecutors.NicknameCommand.executor,
        "naming" -> namingExecutor
      ), whenArgInsufficient = Some(printDescriptionExecutor), whenBranchNotFound = Some(printDescriptionExecutor)
    ).asNonBlockingTabExecutor()
  }

  private object Messages {
    val commandDescription: TargetedEffect[CommandSender] =
      MessageEffect {
        List(
          s"$RED[Usage]",
          s"$RED/mebius naming [name]",
          s"$RED  現在頭に装着中のMEBIUSに[name]を命名します。",
          "",
          s"$RED/mebius nickname",
          s"$RED  MEBIUSから呼ばれる名前を表示します",
          "",
          s"$RED/mebius nickname set [name]",
          s"$RED  MEBIUSから呼ばれる名前を[name]に変更します",
          "",
          s"$RED/mebius nickname reset",
          s"$RED  MEBIUSからの呼び名をプレイヤー名(初期設定)に戻します",
          ""
        )
      }
  }

  object ChildExecutors {
    val printDescriptionExecutor: ContextualExecutor = ContextualExecutorBuilder.beginConfiguration()
      .execution { _ => IO(Messages.commandDescription) }
      .build()

    val namingExecutor: ContextualExecutor = playerCommandBuilder
      .argumentsParsers(List(Parsers.identity))
      .execution { context =>
        // TODO cleanup / do not use Boolean
        def setName(player: Player, name: String): Boolean = {
          val updatedProperty = ItemStackMebiusCodec
            .decodeMebiusProperty(player.getInventory.getHelmet)
            .map {
              _.copy(mebiusName = name)
            }

          updatedProperty.foreach { newProperty =>
            val newDisplayName = ItemStackMebiusCodec.displayNameOfMaterializedItem(newProperty)

            player.sendMessage(s"$newDisplayName${RESET}に命名しました。")
            SeichiAssist.playermap.apply(player.getUniqueId).mebius
              .speakForce(s"わーい、ありがとう！今日から僕は$newDisplayName${RESET}だ！")

            player.getInventory.setHelmet(ItemStackMebiusCodec.materialize(newProperty))
          }

          updatedProperty.nonEmpty
        }

        val newName = s"${context.args.parsed.head.asInstanceOf[String]} ${context.args.yetToBeParsed.mkString(" ")}"

        if (!setName(context.sender, newName)) {
          IO(MessageEffect(s"${RED}命名はMEBIUSを装着して行ってください."))
        } else IO(emptyEffect)
      }
      .build()

    object NicknameCommand {
      // TODO cleanup / do not use Boolean
      private def setNickname(player: Player, name: String): Boolean = {
        val updatedProperty = ItemStackMebiusCodec
          .decodeMebiusProperty(player.getInventory.getHelmet)
          .map {
            _.copy(ownerNickname = Some(name))
          }

        updatedProperty.foreach { newProperty =>
          player.getInventory.setHelmet(ItemStackMebiusCodec.materialize(newProperty))
          SeichiAssist.playermap.apply(player.getUniqueId).mebius
            .speakForce(s"わーい、ありがとう！今日から君のこと$GREEN$name${RESET}って呼ぶね！")
        }

        updatedProperty.nonEmpty
      }

      private val checkNicknameExecutor = playerCommandBuilder
        .execution { context =>
          IO(MessageEffect {
            MebiusListener.getNickname(context.sender)
              .fold {
                s"${RED}呼び名の確認はMEBIUSを装着して行ってください."
              } { name =>
                s"${GREEN}現在のメビウスからの呼び名 : $name"
              }
          })
        }
        .build()

      private val resetNicknameExecutor = playerCommandBuilder
        .execution { context =>
          val message = if (setNickname(context.sender, context.sender.getName)) {
            s"${GREEN}メビウスからの呼び名を${context.sender.getName}にリセットしました."
          } else {
            s"${RED}呼び名のリセットはMEBIUSを装着して行ってください."
          }

          IO(MessageEffect(message))
        }
        .build()

      private val setNicknameExecutor = playerCommandBuilder
        .argumentsParsers(List(Parsers.identity), onMissingArguments = printDescriptionExecutor)
        .execution { context =>
          val newName = s"${context.args.parsed.head.asInstanceOf[String]} ${context.args.yetToBeParsed.mkString(" ")}"
          val message = if (!setNickname(context.sender, newName)) {
            s"${RED}呼び名の設定はMEBIUSを装着して行ってください."
          } else {
            s"${GREEN}メビウスからの呼び名を${newName}にセットしました."
          }

          IO(MessageEffect(message))
        }
        .build()

      val executor: BranchedExecutor = BranchedExecutor(Map(
        "reset" -> resetNicknameExecutor,
        "set" -> setNicknameExecutor
      ), whenArgInsufficient = Some(checkNicknameExecutor), whenBranchNotFound = Some(checkNicknameExecutor))
    }

  }

}
